### ISSUE TITLE
Arrays passed to ccall by pointer are reinterpreted regardless of the element type

### ISSUE THREAD

**Issue by Enet4** [JuliaLang/julia#29850]
I stumbled upon an odd conversion behaviour from arrays to pointers. It is described in [this question on Stack Overflow](https://stackoverflow.com/q/53055762/1233251), but I'll be summarizing the issue here.

```
Julia Version 1.0.1
Commit 0d713926f8 (2018-09-29 19:05 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-6400 CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```

Considering this native function that makes the sum of all even numbers:

```c
uint32_t sum_of_even(const uint32_t *numbers, size_t length);
```

I was defining a wrapper function like this:

```julia
lib = Libdl.dlopen(libname)
sumofeven_sym = Libdl.dlsym(lib, :sum_of_even)

sumofeven(a) = ccall(
    sumofeven_sym,
    UInt32, (Ptr{UInt32}, Csize_t),
    a, length(a))
```

The [documentation on C function calling for bits types](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#man-bits-types-1) suggests that a conversion to an array is not reinterpret cast, so one should be unable to pass an array of some element type to something accepting a pointer to another element type. However, I can actually call `sumofeven` like this:

```julia
sumofeven(Float32[1, 2, 3, 4, 5, 6])
```

... which triggers undefined behaviour. And for reasons which I could not yet grasp, if the function input parameters are modified to accept a `Ref{UInt32}` instead of a `Ptr{UInt32}`, the program correctly prevents me from calling the native function. 

```txt
ERROR: LoadError: MethodError: Cannot `convert` an object of type Array{Float32,1} to an object of type UInt32
Closest candidates are:
  convert(::Type{T<:Number}, !Matched::T<:Number) where T<:Number at number.jl:6
  convert(::Type{T<:Number}, !Matched::Number) where T<:Number at number.jl:7
  convert(::Type{T<:Integer}, !Matched::Ptr) where T<:Integer at pointer.jl:23
  ...
```

I've been told that this might be a bug, hence this report. (I think?) I understand that arguments to `ccall` are automatically converted via `Base.unsafe_convert(Base.cconvert(x))`, and that [`cconvert`](https://docs.julialang.org/en/v1/base/c/#Base.cconvert) should not take an object to produce a `Ptr`, but this seems to contradict with the documentation above.

**Comment by JeffBezanson** [comment]
`Ref{UInt32}` means passing a single UInt32 by reference, not an array.

**Comment by Enet4** [comment]
@JeffBezanson Thanks for the comment (I do have some knowledge gaps in this language). I can't help but think that this makes the documentation even more confusing. The section on [when to use `T`, `Ptr{T}` and `Ref{T}`](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#When-to-use-T,-Ptr{T}-and-Ref{T}-1) actually advises that `Ref{T}` should generally be used for the input argument type when that native function is accepting a pointer, which could be misinterpreted to assume that this also applies to "pointer-based" runtime arrays.

**Comment by yuyichao** [comment]
I don't think this is a bug. `Ptr` is generally the less restrictive version of `Ref` so it will do pointer cast for you. There are no undefined behavior though. The result you get might not be exactly what you want but that's just what you get when you reinterpret integer as floating point.

I do acknowledge that some of what I mentioned are not very well documented though so the doc could be improved.

**Comment by Enet4** [comment]
> The result you get might not be exactly what you want but that's just what you get when you reinterpret integer as floating point.

You're right, this particular example is not necessarily UB, although it could be once we passed an array of smaller types (`UInt8`). There's also an implementation detail in the library that triggers a panic on integer arithmetic overflow, which is not compatible with C, and as such leads also to some form of nasal demons.

>  I do acknowledge that some of what I mentioned are not very well documented though so the doc could be improved.

I believe this would be beneficial. If `Ptr` is meant to be flexible like that, it surprises me to read this in the documentation:

> When an array is passed to C as a `Ptr{T}` argument, it is not reinterpret-cast: Julia requires that the element type of the array matches `T`, and the address of the first element is passed.

**Comment by yuyichao** [comment]
> it is not reinterpret-cast

The `it` here actually refers to the "an array" rather than the "a `Ptr{T}` argument", which I do believe is grammatically correct. Together with the following paragraph

> Therefore, if an `Array` contains data in the wrong format, it will have to be explicitly converted using a call such as `trunc(Int32, a)`.

it should be unambiguous that this implies that passing an array of the wrong format is allowed and it's the content that isn't converted.

I can see how this can be confusing even though it's correct though. The use of "reinterpret" here is especially troubling and I believe it probably better to use "convert". `reinterpret` does not change the bit pattern so it's not extremely clear what it means that a reinterpret-cast doesn't appear.

**Comment by Enet4** [comment]
Interesting, now I see where the "no reinterpret-cast" comes from. :+1:

>  >    Therefore, if an Array contains data in the wrong format, it will have to be explicitly converted using a call such as trunc(Int32, a).
> 
> it should be unambiguous that this implies that passing an array of the wrong format is allowed and it's the content that isn't converted.

I might just add: although I might be missing the context of an experienced Julia programmer, I could interpret this as _"Julia requires that the element type of the array matches `T` and this is statically enforced"_ instead of _"Julia requires the programmer to ensure that the element type of the array matches `T`, otherwise it will eat your laundry"_. The same reasoning could be employed on the subsequent sentence, since [`trunc`](https://docs.julialang.org/en/v1/base/math/#Base.trunc) can do type conversions.

**Comment by hycakir** [comment]
There used to be an error (v0.3, v0.4). That part in the current documentation is identical to the corresponding part in v0.4 documentation. 

The documentation is, IMHO, in stark contrast to the actual behavior. The documentation clearly indicates that there should be an error and even provides some workarounds like using `Ptr{Cvoid}` type for passing a `Float64`, which is not necessary.

I think type checking here would have been useful, easily preventing segfaults that might be overlooked or getting incorrect results.

The current behavior seems to start with [this commit.](https://github.com/JuliaLang/julia/commit/f6f616ce50e40b4337feb79a8bad20519ff75e86#diff-d646a6707a27b7ba7c211a5fa6347321).

**Comment by goretkin** [comment]
This just bit me. I had interpreted 
` Julia requires that the element type of the array matches T`
to mean that Julia would raise an error if I tried to pass `[1, 1, 1]` in the position of an argument of type `Ptr{Cdouble}`. 

I can't imagine any disagreement that the documented behavior and actual behavior are different.

It's a backwards-incompatible change, but can the default behavior please be the safer behavior?

**Comment by chethega** [comment]
The docs are seriously wrong and unclear on this point. I agree that the safe behavior would have been better, but we should at least fix the docs to correctly describe current behavior, e.g.

  * `Array{T,N}`

    When an array is passed to C as a `Ptr{T}` argument, then the existing buffer gets reinterpreted:
    Julia mostly assumes that the element type of the array matches `T`, and the address of the first element 
    is passed:
```
cconvert(T::Type, x) = convert(T, x) # do the conversion eagerly in most cases
cconvert(::Type{<:Ptr}, x) = x # but defer the conversion to Ptr to unsafe_convert
unsafe_convert(::Type{Ptr{S}}, a::AbstractArray{T}) where {S,T} = convert(Ptr{S}, unsafe_convert(Ptr{T}, a))
convert(::Type{Ptr{T}}, p::Ptr) where {T} = bitcast(Ptr{T}, p)
unsafe_convert(::Type{Ptr{T}}, a::Array{T}) where {T} = ccall(:jl_array_ptr, Ptr{T}, (Any,), a)
```
The next paragraph tries to explain the `If an array of eltype `Ptr{T}` is passed as a `Ptr{Ptr{T}}` argument...` methods. I understand neither the paragraph (as written in the docs) nor the actual behavior nor the code leading to the actual behavior. 

I would prefer if we could explain the generic dispatches completely and unambiguously (preferably by just quoting 10 lines of code, like the 5 above), explain whatever string handling magic we do, and refer users to `methods(Base.cconvert)` for further reading.

One of the most important things that must be very clear is when copies are created, and when not, i.e. whether we will see the results of C code mutating args.
